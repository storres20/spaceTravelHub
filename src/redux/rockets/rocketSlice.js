import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

export const getRockets = createAsyncThunk('rocket/getRockets', async () => fetch('https://api.spacexdata.com/v3/rockets')
  .then((res) => res.json())
  .then((data) => {
    const rockets = data.map((rocket) => ({
      id: rocket.id,
      rocket_name: rocket.rocket_name,
      description: rocket.description,
      flickr_images: rocket.flickr_images[0],
    }));
    return rockets;
  })
  .catch((error) => {
    throw new Error(error);
  }));

const rocketSlice = createSlice({
  name: 'rocket',
  initialState: { rockets: [], loading: false, refresh: true },
  extraReducers: {
    [getRockets.pending]: (state) => ({ ...state, isloading: true }),
    [getRockets.fulfilled]: (state, action) => ({
      ...state,
      rockets: action.payload,
      isloading: false,
    }),
    [getRockets.rejected]: (state) => ({ ...state, isloading: false }),
  },
});
export default rocketSlice.reducer;
