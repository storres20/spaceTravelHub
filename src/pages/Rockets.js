import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import '../App.css';
import Col from 'react-bootstrap/Col';
import Card from 'react-bootstrap/Card';
import CardGroup from 'react-bootstrap/CardGroup';
import Button from 'react-bootstrap/Button';
import { reserveRocket } from '../redux/rockets/rocketSlice';

function Rockets() {
  const { rockets } = useSelector((state) => state.rockets);
  const dispatch = useDispatch();

  const handleReserve = (id) => {
    dispatch(reserveRocket(id));
  };

  return (
    <CardGroup className="d-grid" style={{ marginTop: '2%', padding: '10px', borderStyle: 'none' }}>
      {rockets.map(({
        id,
        rocketName,
        description,
        flickrImages,
      }) => (
        <Card key={id} className="d-flex flex-sm-column flex-md-row" style={{ marginBottom: '2%', borderStyle: 'none' }}>
          <Col xs={6} md={3}>
            <Card.Img src={flickrImages} />
          </Col>
          <Col xs={12} md={9}>
            <Card.Body>
              <Card.Title>{rocketName}</Card.Title>
              <Card.Text>
                {description}
              </Card.Text>
              <Button onClick={() => handleReserve(id)} variant="primary">Reserve Rocket</Button>
            </Card.Body>
          </Col>
        </Card>
      ))}
    </CardGroup>
  );
}

export default Rockets;
