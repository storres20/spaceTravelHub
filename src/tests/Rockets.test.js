import React from 'react';
import renderer from 'react-test-renderer';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import Rockets from '../pages/Rockets';
// import moviesReducer from '../redux/movies/movies';
import rockets from '../redux/rockets/rocketSlice';

// configureStore
const createTestStore = () => {
  const store = configureStore({
    reducer: {
      rockets,
    },
  });

  return store;
};
export default createTestStore;

// Test
describe('My Connected React-Redux Component', () => {
  let store;
  let component;

  beforeEach(() => {
    store = createTestStore();

    // Create a redux store
    component = renderer.create(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    ).toJSON();
  });

  test('renders correctly', () => {
    expect(component).toMatchSnapshot();
  });
});
