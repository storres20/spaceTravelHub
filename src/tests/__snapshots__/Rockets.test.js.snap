// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Connected React-Redux Component renders correctly 1`] = `
<div
  className="d-grid card-group"
  style={
    Object {
      "borderStyle": "none",
      "marginTop": "2%",
      "padding": "10px",
    }
  }
/>
`;
